### PR TITLE
Add cross modal example

### PR DESCRIPTION
**Pull request by JoanFM**
Add cross modal examples

**Work to do**
- [x] Create documents to build Docker Images

- [x] Make models work with batch

- [x] Create structure for `flow-index.yml` and `flow-query.yml`

- [x] Create pods yaml descriptions

- [x] Decide how documents will be input to the system

- [x] Create app.py 

- [x] Create documentation or decide how to provide data

**Post work**
@bwanglzu could you take care of the steps that are not marked?
- [x] Test the results of man_piercing (see their embeddings) and their captions to match research. The image one will not exactly match because we do not resze and crop but just resize

- [ ] Have nice printing and showing of results

- [ ] Put nice (nice results) examples in README

- [ ] Test everything

- [x] See if Docker Images can be made smaller

**Nice to have**

- [ ] Easy executable that indexes data, and queries `test` dataset and runs `evaluation` to reproduce results in paper

### CODE REVIEW

File: cross-modal-search/txt_emb/VSETextEncoder.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        captions = []
+        for sentence in text:
+            tokens = nltk.tokenize.word_tokenize(str(sentence).lower())
+            caption = []
+            caption.append(self.vocab('<start>'))
+            caption.extend([self.vocab(token) for token in tokens])
+            caption.append(self.vocab('<end>'))
+            target = torch.Tensor(caption)
+            captions.append(target)
+        captions.sort(key=lambda x: len(x), reverse=True)
```
Comment (JoanFM) [inline cross-modal-search/txt_emb/VSETextEncoder.py:70]: `VSE` text encoder wants captions in the `batch` sorted by length (from largest to shortest)

File: cross-modal-search/txt_emb/VSETextEncoder.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        lengths = [len(x) for x in captions]
+        targets = torch.zeros(len(captions), max(lengths)).long()
```
Comment (JoanFM) [inline cross-modal-search/txt_emb/VSETextEncoder.py:72]: All `captions` tokenized vectors must be `zero padded` to the longest in the `batch`

**Comment by bwanglzu** [comment]
Build docker images:

- [ ]  `docker build -f Dockerfile -t jinaai/hub.executors.encoders.nlp.vse:latest` 
- [ ]  `docker build -f Dockerfile -t jinaai/hub.executors.encoders.image.vse:latest .`
- [ ]  `python app.py --task=index`

File: cross-modal-search/app.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+        f = Flow().load_config('flow-query.yml')
+        with f:
+            while True:
+                image_file_path = input('please type an image file path to search for a caption: ')
+                if not image_file_path:
+                    break
+                f.search(input_fn=input_search_image_file(image_file_path),
+                         output_fn=lambda x: print_top_k(x, text), top_k=top_k)
+    elif task == 'query-restful':
+        # not working, missing a way to send modality via REST API
```
Comment (JoanFM) [inline cross-modal-search/app.py:113]: Remove this comment and suggest using `jinabox` in a message

File: cross-modal-search/README.md [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+### Index 
+
+Index is run with the following command, where batch_size can be chosen by the user. Index will index both images and captions
+
+```bash
+python app.py -t index -n $num_docs -b $batch_size -d 'f8k'
+```
+
+Not that num_docs should be 8k or 30k depending on the flickr dataset you use. If you decide to index the complete datasets,
```
Comment (rutujasurve94) [inline cross-modal-search/README.md:115]: Note*

File: cross-modal-search/img_emb/__init__.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+__license__ = "Apache-2.0"
+
+import torch
+from torch.autograd import Variable
+# model is a file from the vsepp github
+import sys
+sys.path.append(".")
+from model import VSE
+
+from jina.hub.encoders.image.torchvision import ImageTorchEncoder
```
Comment (JoanFM) [inline cross-modal-search/img_emb/__init__.py:11]: We should remove this dependency

**Comment by JoanFM** [comment]
I think we should fix `jina version to 0.5.0`

**Pull request merged**